### PR TITLE
internaltypes: Helper to ensure no property name is longer than 80 characters

### PR DESCRIPTION
**Pull request by draychev**
Adding a function named `governor()`, which ensures that strings longer than 80 characters are converted to SHA256 hashes of themselves.

Here is an example of the issue this would solve:
```
W0510 16:30:41.323287    3944 controller.go:110] unable to send CreateOrUpdate request, error [network.ApplicationGatewaysClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidResourceName" Message="Resource name k8s-ag-ingress-default-liveliness-test-service-80-bp-8080-test-health-probes-ingress is invalid. The name can be up to 80 characters long. It must begin with a word character, and it must end with a word character or with '_'. The name may contain word characters or '.', '-', '_'." Details=[]]
```

The error above indicates that `k8s-ag-ingress-default-liveliness-test-service-80-bp-8080-test-health-probes-ingress`, which is 84 characters long, is too long.  The newly added code would convert this into `prop-87fc06ba019c10dbf2ae685654c58528a64f95a768b97d3be76e706b7283e341`, which is shorter.

I have not added the `governor()` function to all Name generators.  Once we agree this is an acceptable approach I will do so in a follow-up PR.

This PR here is the first step to fixing https://github.com/Azure/application-gateway-kubernetes-ingress/issues/125

### CODE REVIEW

**Comment by draychev** [comment]
> I think this looks good as such.
> 
> How about taking first 36 char sub-string of name and appending it with 64 (or less) char hash to improve readability ?

@akshaysngupta Subjectively - once we get to 80 characters and above, readability is already compromised.  But I agree that preserving as much as possible from the original name would be helpful.  It'd would be even better if we removed the prefix (coming soon).

File: pkg/appgw/internaltypes.go [hunk 1]
```diff
@@ -76,7 +89,7 @@ func generateHTTPSettingsName(serviceName string, servicePort string, backendPor
 }
 
 func generateProbeName(serviceName string, servicePort string, ingress string) string {
-	return fmt.Sprintf("%s-%v-%v-pb-%s", agPrefix, serviceName, servicePort, ingress)
+	return governor(fmt.Sprintf("%s-%v-%v-pb-%s", agPrefix, serviceName, servicePort, ingress))
```
Comment (akshaysngupta) [inline pkg/appgw/internaltypes.go]: we should add a log mentioning that we shortened the name for this ingress/rule/path/backend(service)

Comment (draychev) [reply pkg/appgw/internaltypes.go]: SGTM
Here is a sample log line: `•••••••••••I0514 12:27:22.761899   30440 internaltypes.go:66] Prop name xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx with length 81 is longer than 80; Transformed to xxxxxxxxxxxxxxx-4d1f7bb876c6aae013f06a7430b8c6545ff30d8f9252838eb18b6357c1a2ba13`

Comment (akshaysngupta) [reply pkg/appgw/internaltypes.go]: Having a shorter hash like 128-bit and removing the ingress prefix, this would look better.

Comment (draychev) [reply pkg/appgw/internaltypes.go]: I was actually thinking about using MD5 -- https://github.com/Azure/application-gateway-kubernetes-ingress/pull/128#pullrequestreview-236403946
Agreed - no need for a crytpographically strong hash - as long as we have a deterministic hash with some  fairly low probability of collision - I think MD5 here is perfect. 32 characters for the hash, 1 for the separator, 47 for the human readable part.

**Pull request merged**